// biz9-app/main.js
/*
Copyright 2023 Certified CoderZ
Author: certifiedcoderz@gmail.com (Certified CoderZ)
License GNU General Public License v3.0
Description: BiZ9 Framework: App - Main
*/
const path = require('path')
const async = require('async')
const moment = require('moment')
const biz9_config_file = require(path.join(__dirname, '../../biz9_config.js'));
const {get_title_url,get_date_time_pretty,get_date_str,get_date_time_str,get_month_title_short,get_date_time_obj,get_time_str} = require(process.env.BIZ9_HOME+'/biz9-utility/src/code/index.js')
const get_new_item_main = (data_type,tbl_id) => {
    if(!tbl_id){
        tbl_id=0;
    }
    return {data_type:data_type,tbl_id:tbl_id};
}
const set_biz_item_main = (item,options) => {
    //options = get_photo, get_mp3, get_date, get_count
    return new Promise((callback) => {
        let error = null;
        async.series([
            function(call) {
                if(!options){
                    options={};
                }
                call();
            },
            function(call){
                if(options.get_count){
                    if(!item.view_count){
                        item.view_count='0';
                    }
                    if(!item.like_count){
                        item.like_count='0';
                    }
                    if(!item.review_count){
                        item.review_count='0';
                    }
                }
                call();
            },
            function(call) {
                if(options.get_mp3){
                    if(item.mp3filename){
                        item.mp3_url = biz9_config_file.FILE_URL+item.mp3filename;
                    }else{
                        item.mp3_url=null;
                    }
                }
                call();
            },
            function(call) {
                if(options.get_photo){
                    let  no_photo_str='/images/no_image.png';
                    let _photo_size_album='';
                    let _photo_size_thumb='thumb_size_';
                    let _photo_size_mid='mid_size_';
                    let _photo_size_large='large_size_';
                    let _photo_size_square_thumb='square_thumb_size_';
                    let _photo_size_square_mid='square_mid_size_';
                    let _photo_size_square_large='square_large_size_';
                    if(!item.photofilename){
                        item.photofilename=null;
                    }
                    item.photo_obj={
                        album_url : (item.photofilename) ? biz9_config_file.FILE_URL+item.photofilename : no_photo_str,
                        thumb_url : (item.photofilename) ?biz9_config_file.FILE_URL+_photo_size_thumb+item.photofilename : no_photo_str,
                        mid_url   : (item.photofilename) ?biz9_config_file.FILE_URL+_photo_size_mid+item.photofilename : no_photo_str,
                        large_url : (item.photofilename) ?biz9_config_file.FILE_URL+_photo_size_large+item.photofilename : no_photo_str,
                        thumb_url : (item.photofilename) ?biz9_config_file.FILE_URL+_photo_size_thumb+item.photofilename : no_photo_str,
                        square_thumb_url : (item.photofilename) ? biz9_config_file.FILE_URL+_photo_size_square_thumb+item.photofilename : no_photo_str,
                        square_mid_url   : (item.photofilename) ? biz9_config_file.FILE_URL+_photo_size_square_thumb+item.photofilename : no_photo_str,
                        square_large_url : (item.photofilename) ? biz9_config_file.FILE_URL+_photo_size_square_thumb+item.photofilename : no_photo_str,
                    };
                }
                call();
            },
            function(call){
                if(options.get_date){
                    no_date_str='';
                    if(!item.date_create){
                        item.date_create= new moment().toISOString();
                    }
                    item.date_obj={
                        pretty_create: (item.date_create) ? get_date_time_pretty(item.date_create) : no_date_str,
                        pretty_update: (item.date_create) ? get_date_time_pretty(item.date_save): no_date_str,
                        full_date_create: (item.date_create) ? get_date_str(item.date_create) : no_date_str,
                        full_date_update: (item.date_create) ? get_date_str(item.date_save) : no_date_str,
                        full_date_time_create: (item.date_create) ? get_date_time_str(item.date_create) : no_date_str,
                        full_date_time_update: (item.date_create) ? get_date_time_str(item.date_save) : no_date_str,
                        month_create: (item.date_create) ? get_month_title_short(1+get_date_time_obj(item.date_create).month()) : no_date_str,
                        month_update: (item.date_create) ? get_month_title_short(1+get_date_time_obj(item.date_save).month()) : no_date_str,
                        mo_create: (item.date_create) ? (1+get_date_time_obj(item.date_create).month()) : no_date_str,
                        mo_update: (item.date_create) ? (1+get_date_time_obj(item.date_save).month()) : no_date_str,
                        date_create: (item.date_create) ? get_date_time_obj(item.date_create).date() : no_date_str,
                        year_create: (item.date_create) ? get_date_time_obj(item.date_create).year() : no_date_str,
                        year_update: (item.date_create) ? get_date_time_obj(item.date_save).year() : no_date_str,
                        time_create: (item.date_create) ? get_time_str(item.date_create) : no_date_str,
                        time_update: (item.date_create) ? get_time_str(item.date_save) : no_date_str,
                    }
                }
                call();
            },
        ]).then(result => {
            callback([error,item]);
        }).catch(err => {
            console.error("--Error-App-Main-Set-Biz-Item-Main-Error--",err);
        });
    });
}
const set_biz_item_old = (item) => {
    let  no_photo_str='/images/no_image.png';
    let _photo_size_album='';
    let _photo_size_thumb='thumb_size_';
    let _photo_size_mid='mid_size_';
    let _photo_size_large='large_size_';
    let _photo_size_square_thumb='square_thumb_size_';
    let _photo_size_square_mid='square_mid_size_';
    let _photo_size_square_large='square_large_size_';
    if(item.mp3filename){
        item.mp3_url = biz9_config_file.FILE_URL+item.mp3filename;
    }
    if(!item.view_count){
        item.view_count='0';
    }
    if(!item.like_count){
        item.like_count='0';
    }
    if(!item.photofilename){
        item.photofilename=null;
    }
    item.photo_obj={
        album_url : (item.photofilename) ? biz9_config_file.FILE_URL+item.photofilename : no_photo_str,
        thumb_url : (item.photofilename) ?biz9_config_file.FILE_URL+_photo_size_thumb+item.photofilename : no_photo_str,
        mid_url   : (item.photofilename) ?biz9_config_file.FILE_URL+_photo_size_mid+item.photofilename : no_photo_str,
        large_url : (item.photofilename) ?biz9_config_file.FILE_URL+_photo_size_large+item.photofilename : no_photo_str,
        thumb_url : (item.photofilename) ?biz9_config_file.FILE_URL+_photo_size_thumb+item.photofilename : no_photo_str,
        square_thumb_url : (item.photofilename) ? biz9_config_file.FILE_URL+_photo_size_square_thumb+item.photofilename : no_photo_str,
        square_mid_url   : (item.photofilename) ? biz9_config_file.FILE_URL+_photo_size_square_thumb+item.photofilename : no_photo_str,
        square_large_url : (item.photofilename) ? biz9_config_file.FILE_URL+_photo_size_square_thumb+item.photofilename : no_photo_str,
    };
    if(!item.review_count){
        item.review_count='0';
    }
    if(!item.view_count){
        item.view_count='0';
    }
    no_date_str='';
    if(!item.date_create){
        item.date_create= new moment().toISOString();
    }
    item.date_obj={
        pretty_create: (item.date_create) ? get_date_time_pretty(item.date_create) : no_date_str,
        pretty_update: (item.date_create) ? get_date_time_pretty(item.date_save): no_date_str,
        full_date_create: (item.date_create) ? get_date_str(item.date_create) : no_date_str,
        full_date_update: (item.date_create) ? get_date_str(item.date_save) : no_date_str,
        full_date_time_create: (item.date_create) ? get_date_time_str(item.date_create) : no_date_str,
        full_date_time_update: (item.date_create) ? get_date_time_str(item.date_save) : no_date_str,
        month_create: (item.date_create) ? get_month_title_short(1+get_date_time_obj(item.date_create).month()) : no_date_str,
        month_update: (item.date_create) ? get_month_title_short(1+biz9.get_date_time_obj(item.date_save).month()) : no_date_str,
        mo_create: (item.date_create) ? (1+get_date_time_obj(item.date_create).month()) : no_date_str,
        mo_update: (item.date_create) ? (1+get_date_time_obj(item.date_save).month()) : no_date_str,
        date_create: (item.date_create) ? get_date_time_obj(item.date_create).date() : no_date_str,
        year_create: (item.date_create) ? get_date_time_obj(item.date_create).year() : no_date_str,
        year_update: (item.date_create) ? get_date_time_obj(item.date_save).year() : no_date_str,
        time_create: (item.date_create) ? get_time_str(item.date_create) : no_date_str,
        time_update: (item.date_create) ? get_time_str(item.date_save) : no_date_str,
    }
    if(biz9_config_file.BIZ_MAP==true){
        if(item.field_1 || item.date_1){
            for(let a=0;a<21;a++){
                if(item['field_'+a]){
                    item[item['field_'+a]]=item['value_'+a] ? (item['value_'+a]) : '';
                }
                if(item['date_'+a]){
                    item[item['date_'+a]]= get_date_time_obj(item['date_value_'+a]) ? (item['date_value_'+a]) : '';
                }
            }
        }
    }
    for (const key in item) {
        if(item[key] != 'date_create' || 'photofilename'){
            if(!item[key]){
                item[key]='';
            }
        }
    }
    return item;
}
module.exports = {
    get_new_item_main,
    set_biz_item_main
};
